### ISSUE TITLE
Build fails with oneDNN, missing CMakeLists.txt

### ISSUE THREAD

**Issue by AlbertoGP** [charles-r-earp/autograph#22]
Downloaded as zip file, ran `cargo run --example mnist_dense --features "datasets" --release`:

`CMake Error: The source directory "autograph-master/oneDNN" does not appear to contain CMakeLists.txt.`

This was just a quick test, maybe I’m missing some obvious step.

**Comment by charles-r-earp** [comment]
The oneDNN directory is a git submodule. If you have git installed, you should be able to run git submodule update --init --recursive. You can also download oneDNN https://github.com/oneapi-src/oneDNN and copy it manually.

**Comment by AlbertoGP** [comment]
I see, thanks. I’ve done it, and it turns out that OpenMP is also required: I know it would be very slow without it, but from the README (*[Supported platforms](https://github.com/AlbertoGP/autograph#supported-platforms)*) and the following warning message from cmake it seems that it should be optional, however it is a fatal error:
```
   Compiling autograph v0.0.1 (/home/alberto/stor/src/autograph)
error: failed to run custom build command for `autograph v0.0.1 (/home/alberto/stor/src/autograph)`

Caused by:
  process didn't exit successfully: `/home/alberto/stor/src/autograph/target/debug/build/autograph-846f68cbfbb655e0/build-script-build` (exit code: 101)
--- stdout
...
--- stderr
CMake Error at cmake/OpenMP.cmake:95 (message):
  OpenMP library could not be found.  Proceeding might lead to highly
  sub-optimal performance.
Call Stack (most recent call first):
  CMakeLists.txt:114 (include)


thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

build script failed, must exit now', /home/alberto/.cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.44/src/lib.rs:885:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

warning: build failed, waiting for other jobs to finish...
error: build failed
```

I don’t care for now about performance, just want to see it run with as little extra installation and configuration as possible.

I’m adding the extra installation steps to the README, and will submit a pull request one it’s working.
For now you can find it here: https://github.com/AlbertoGP/autograph#getting-started

**Comment by charles-r-earp** [comment]
Thanks for this. Yes, I knew about this issue. The oneDNN library can work without OpenMP, but if it does find it then the library must be linked to OpenMP or else the examples (and any dependent targets) fail at link time. So right now the build script only links to OpenMP on Linux. 
It should be possible to extract the required information (whether and where oneDNN is getting OpenMP) and that would fix this. CMake generates a CMakeCache.txt which has this information (similar to the output of running cmake). If you run:
```
cargo clean && cargo build -vv // very verbose
``
It should dump the cmake build output, which will tell you what version of OpenMP is found among other things. 
It was my intent to bundle the bare minimum libraries in the crate so that it could be used without extra installation steps. Note that AFAIK, cargo runs the git clone --recursive command, so it should work when added as a dependency (whether from crates.io or github). If you clone from the browser, and don't have git installed, then yes you have to copy the deps manually.

So in short, it would be nice to ensure that it will work on all platforms with or without OpenMP. If necessary, we could add a environmental variable for the path to OpenMP, if FindOpenMP.cmake fails (which it apparently does on macOS).

**Comment by AlbertoGP** [comment]
Ok, the issue was just that I had to install, instead of `libomp` (clang), `libgomp` (gcc).

oneDNN stil doesn’t compile as-is because I’m doing this on an IBM POWER9 CPU and the compiler does not have the `-msse4.1` option, but that’s a patch for oneDNN (https://github.com/oneapi-src/oneDNN/pull/767).

After that’s fixed, the build script does not find `libdnnl.a` because of #23, but that’s an easy fix (#24).
